### PR TITLE
Reorganize Rules, Matchers, and Helpers

### PR DESCRIPTION
**Pull request by jacknagz**
to: @ryandeivert 
cc: @airbnb/streamalert-maintainers @zbuc
size: small

## Background
Currently, helpers, matchers, and rules are all nested within the `rules/` directory.  This is slightly confusing and also makes importing tricky.  This change moves all pieces into their own respective folder

## Changes
* Move matchers and helpers into their own folders
* Recursively import matchers and rules
* Update documentation

### CODE REVIEW

**Comment by jacknagz** [comment]
re coveralls: @ryandeivert should we move our rules testing framework into `nose` to recognize the coverage?

**Comment by ryandeivert** [comment]
@jacknagz I'm not sure I understand what you mean. `nose` and `coveralls` are really only for unit tests, they have no understanding of integration or functional tests.

Are you concerned about bits of code that may be duplicative? If so, you can do something similar to what I did with `test/unit/stream_alert_alert_processor/test_outputs.py`.. create helper functions in `stream_alert_cli/helpers.py` and import with:
```
from stream_alert_cli.helpers import (
    create_lambda_function,
    put_mock_creds
)
```

If that's not what you're getting at, let's talk more offline.

**Comment by ryandeivert** [comment]
@jacknagz also RE your comment about increasing coverage, this is super helpful from the repo root:

`./test/scripts/unit_tests.sh; coverage html; open htmlcov/index.html`

Use the code block highlighting to see what's covered/missed. V nice.

File: stream_alert/rule_processor/main.py [hunk 1]
```diff
@@ -18,25 +18,27 @@
...
-                                      not x.endswith('.pyc') and
-                                      not x.startswith('__init__'), files)
-    for import_file in filtered_files:
-        package_path = root.replace('/', '.')
-        import_module = os.path.splitext(import_file)[0]
-        modules_to_import.add('{}.{}'.format(package_path, import_module))
-
+for folder in ('matchers/', 'rules/'):
+    for root, dirs, files in os.walk(folder):
+        filtered_files = filter(lambda x: not x.startswith('.') and
```
Comment (ryandeivert) [inline stream_alert/rule_processor/main.py]: I know this isn't new code, but didn't you say pylint doesn't like `filter` anymore? Along with `map`, etc? Thinking forward with adding `pylint` to CI, we should probably knock these out as we see them and switch to list/dict comprehension.

File: stream_alert/rule_processor/main.py [hunk 2]
```diff
@@ -18,25 +18,27 @@
...
+                                          not x.endswith('.pyc') and
+                                          not x.startswith('__init__'),
+                                          files)
+        for import_file in filtered_files:
+            # Greater than two because that indicates there's only one folder
+            if len(root.split('/')) > 2:
```
Comment (ryandeivert) [inline stream_alert/rule_processor/main.py]: You could avoid splitting this string into a list temporarily (which creates a new object in memory) by using `if root.count('/') > 1:...`. This should apply the same logic you're looking for.

File: stream_alert/rule_processor/main.py [hunk 3]
```diff
@@ -18,25 +18,27 @@
...
+                package_path = root.replace('/', '.')
+            else:
+                package_path = root.rstrip('/')
+            import_module = os.path.splitext(import_file)[0]
+            print package_path, import_module
```
Comment (ryandeivert) [inline stream_alert/rule_processor/main.py]: Are these some rogue `print` statements that were left behind from testing? Or do we want to keep them here? There's this one and one on L41 below.

File: stream_alert/rule_processor/main.py (hunk 1 above)
Comment (jacknagz) [inline stream_alert/rule_processor/main.py]: That's fair, I'll convert it (even though the ruby programmer in me really loves filter and map)

File: stream_alert/rule_processor/main.py (hunk 2 above)
Comment (jacknagz) [inline stream_alert/rule_processor/main.py]: Good tip, I'll make that change

File: stream_alert/rule_processor/main.py (hunk 3 above)
Comment (jacknagz) [inline stream_alert/rule_processor/main.py]: This is why you don't develop late on Fridays 🤣

**Pull request merged**